src/ClientHandle.cpp

```diff
@@ -788,12 +789,39 @@ void cClientHandle::HandleEnchantItem(UInt8 a_WindowID, UInt8 a_Enchantment)
 
 	if (Item.EnchantByXPLevels(BaseEnchantmentLevel))
 	{
-		if (m_Player->IsGameModeCreative() || m_Player->DeltaExperience(-m_Player->XpForLevel(BaseEnchantmentLevel)) >= 0)
+		const bool IsCreative = m_Player->IsGameModeCreative();
+		const auto XpRequired = m_Player->XpForLevel(BaseEnchantmentLevel);
+		cItem LapisStack = *Window->m_SlotArea->GetSlot(1, *m_Player);  // Make a copy of the lapis stack
+		const auto LapisRequired = a_Enchantment+1;
+
+		// Only allow enchantment if the player is creative or has sufficient levels and lapis to enchant
+		if (IsCreative || ((m_Player->GetCurrentXp() >= XpRequired) && LapisStack.GetCount() >= LapisRequired))
 		{
+			// If they weren't creative then subtract xp and lapis
+			if (!IsCreative)
+			{
+				/** We need to reduce the player's level by the amount of lapis required
+				However we need to keep the resulting percentage filled the same */
+				// Get current level and fill percent
+				const auto CurrentLevel = m_Player->GetXpLevel();
+				const auto TargetLevel = CurrentLevel - LapisRequired;
+				const auto CurrentFillPercent = m_Player->GetXpPercentage();
+
+				// Remove current xp and add amount for target level
+				const auto DeltaForLevel = -m_Player->GetCurrentXp() + m_Player->XpForLevel(TargetLevel);
+				// Add correct amount to get the same fill percent
+				const auto DeltaForPercent = CurrentFillPercent*(m_Player->XpForLevel(TargetLevel + 1) - m_Player->XpForLevel(TargetLevel));
+				m_Player->DeltaExperience(DeltaForLevel + DeltaForPercent);
+
+				// Now reduce the lapis in our stack and send it back
+				LapisStack.AddCount(-LapisRequired);
+				Window->m_SlotArea->SetSlot(1, *m_Player, LapisStack);
+			}
+			// Set the item slot to our new enchanted item
 			Window->m_SlotArea->SetSlot(0, *m_Player, Item);
-			Window->SendSlot(*m_Player, Window->m_SlotArea, 0);
 			Window->BroadcastWholeWindow();
 
+			// Remove enchantment choices
 			Window->SetProperty(0, 0, *m_Player);
```

Comment:
The properties are used to send the possible enchantment to the player
see here: https://wiki.vg/index.php?title=Protocol&oldid=14204#Window_Property